chainerrl/agents/dqfd.py

```diff
@@ -0,0 +1,620 @@
+                                   target_update_interval, clip_delta,
+                                   phi, target_update_method, soft_update_tau,
+                                   n_times_update, average_q_decay,
+                                   average_loss_decay, batch_accumulator,
+                                   episodic_update, episodic_update_len,
+                                   logger, batch_states)
+
+        self.minibatch_size = minibatch_size
+        self.n_pretrain_steps = n_pretrain_steps
+        self.demo_supervised_margin = demo_supervised_margin
+        self.loss_coeff_supervised = loss_coeff_supervised
+        self.loss_coeff_l2 = loss_coeff_l2
+        self.loss_coeff_nstep = loss_coeff_nstep
+        self.bonus_priority_demo = bonus_priority_demo
+        self.bonus_priority_agent = bonus_priority_agent
+
+        self.average_loss_1step = 0
+        self.average_loss_nstep = 0
+        self.average_loss_supervised = 0
+
+        self.optimizer.add_hook(
+            chainer.optimizer_hooks.WeightDecay(loss_coeff_l2))
+
+        # Overwrite DQN's replay updater.
+        # TODO: Is there a better way to do this?
+        self.replay_updater = DemoReplayUpdater(
+            replay_buffer=self.replay_buffer,
+            update_func=self.update,
+            batchsize=minibatch_size,
+            episodic_update=episodic_update,
+            episodic_update_len=episodic_update_len,
+            n_times_update=n_times_update,
+            replay_start_size=replay_start_size,
+            update_interval=update_interval,
+        )
+
+        # TODO: Should this really go here? Move into train function?
+        self.pretrain()
+
+    def pretrain(self):
+        """Uses purely expert demonstrations to do pre-training
+        """
+        logger = getLogger(__name__)
+        for tpre in range(self.n_pretrain_steps):
+            self.replay_updater.update_from_demonstrations()
+            if tpre % self.target_update_interval == 0:
+                logger.info('PRETRAIN-step:%s statistics:%s',
+                            tpre, self.get_statistics())
+                self.sync_target_network()
+
+    def update(self, experiences_agent, experiences_demo):
+        """Combined DQfD loss function for Demonstration and agent/RL.
+        """
+        num_exp_agent = len(experiences_agent)
+        experiences = experiences_agent + experiences_demo
+        exp_batch = batch_experiences(experiences, xp=self.xp, phi=self.phi,
+                                      gamma=self.gamma,
+                                      batch_states=self.batch_states)
+
+        exp_batch['weights'] = self.xp.asarray(
+            [elem[0]['weight']for elem in experiences], dtype=self.xp.float32)
+
+        errors_out = []
+        loss_q_nstep, loss_q_1step = self._compute_ddqn_losses(
+            exp_batch, errors_out=errors_out)
+
+        # Add the agent/demonstration bonus priorities and update
```

Comment:
Should we be updating the priorities even during pretraining? Because it looks like that's what's happening.

Reply:
This is unclear from the paper. After some discussion, we decided it makes more sense to not have it update during pre-training. I'll make this change.